### PR TITLE
Add array implementation of Context.

### PR DESCRIPTION
**Pull request by anuraaga**
Benchmarks: https://gist.github.com/anuraaga/db897e6121f9e01b4c5ec5f3de60a60f

Like before, we see the write performance of array-based to be better at all sizes (`writeAll` is what matters), and `read` for the worst case of `readLast` does about the same even with 10 elements, and gets a little worse at 20. Given write has to copy the whole array, this seems to show that arrays in Java are pretty well optimized.

I think removing the complex class is worth it, but open to any thoughts. For the future, we could go further by defining `OneElementContext`, `TwoElementContext`, etc to optimize the vastly common cases of ~5 elements, and could probably also consider just using a simple `HashMap` if it gets big if there are concerns about the Array's asymptotes (though even with 40 elements, it seems not that bad).

### CODE REVIEW

**Review by breedx-splk — COMMENTED** [review]
I added my $0.02.  I also peeked at the benchmark gist -- and maybe I'm misreading it but doesn't this show a worse performance? 
```
ContextBenchmark.writeAll                                   40  thrpt   15     650.433 ±    34.787  ops/ms
ContextBenchmark.writeAll                                   40  thrpt   15     856.471 ±    10.210  ops/ms
```
Doesn't it show that the top one (trie) doing ~35 operations/ms and the new array-based one doing ~10 operations/ms...and isn't more operations better?

**Comment by jkwatson** [comment]
do you have benchmarks on memory usage of the two implementations as well? Not sure how important it is, but I think it would be an interesting point to consider when choosing which we want.

**Review by anuraaga — COMMENTED** [review]
> Doesn't it show that the top one (trie) doing ~35 operations/ms and the new array-based one doing ~10 operations/ms...and isn't more operations better?

The 35 / 10 are the standard deviations, so the error range, the averages are next to it 650 / 856. Generally when comparing changes, I add the standard deviation to the old value and subtract from the new one to give a worst-case comparison, but we're doing pretty well here.

> do you have benchmarks on memory usage of the two implementations as well? Not sure how important it is, but I think it would be an interesting point to consider when choosing which we want.

We can see allocation rate in the benchmarks at least, it should be an OK approximation of memory usage, especially for array implementation.

Trie
```
ContextBenchmark.writeAll:·gc.alloc.rate.norm                5  thrpt   15     344.000 ±     0.001    B/op
ContextBenchmark.writeAll:·gc.alloc.rate.norm               20  thrpt   15    2194.667 ±    15.047    B/op
ContextBenchmark.writeAll:·gc.alloc.rate.norm               40  thrpt   15    5696.001 ±    50.597    B/op
```

Array
```
ContextBenchmark.writeAll:·gc.alloc.rate.norm                5  thrpt   15     280.000 ±     0.001    B/op
ContextBenchmark.writeAll:·gc.alloc.rate.norm               20  thrpt   15    2320.000 ±     0.001    B/op
ContextBenchmark.writeAll:·gc.alloc.rate.norm               40  thrpt   15    7840.000 ±     0.001    B/op
```

For 20 elements it seems about even, definitely for 40 the copy of the entire array starts to manifest itself. For 5 elements, my general target size because even if the user adds a couple of items to context, that'll be the general pattern, array seems to be using less memory. I'm not sure how the persistence is being reflected in memory usage here, but I guess one nice thing about the array is if an old context is gone, we know all the backing storage can be GC'd, with the persistent trie it's a bit hard to reason about due to the compaction, presumably some but not all does get retained even when a parent context gets collected.

File: context/src/main/java/io/opentelemetry/context/ArrayBasedContext.java [hunk 1]
```diff
@@ -42,5 +45,40 @@ static Context root() {
...
+      if (entries[i] == key) {
+        if (entries[i + 1] == value) {
+          return this;
+        }
+        Object[] newEntries = entries.clone();
+        newEntries[i + 1] = value;
+        return new ArrayBasedContext(newEntries);
+      }
+    }
+    Object[] newEntries = Arrays.copyOf(entries, entries.length + 2);
```
[the remark this answers is not in the document]
Comment (breedx-splk) [reply context/src/main/java/io/opentelemetry/context/ArrayBasedContext.java:79]: Ah good point!  👍 I overlooked that.

**Comment by breedx-splk** [comment]
Thanks for clearing me up on the stdev thing -- the formatting threw me off, but yeah, I can see now that it's very much `x ± y units`, and I was only looking at `y`.

**Review by jkwatson — APPROVED** [review]
This seems much much simpler, so I approve, given the benchmarks.

**Pull request merged**